### DISCUSSION TITLE
菜单问题

### DISCUSSION THREAD

**Discussion by 204065248** [Milkdown/milkdown#805] [Q&A] [answered]
我想在菜单中实现一个下拉框表情，但是不知道如何实现。因为他的类型被限制为SelectConfig或者ButtonConfig了，请问如何进行拓展呢，比如我想使用antd的dropdown组件

**Comment by Saul-Mirone** [comment, marked as answer]
It's not possible for now, but will be supported in the future :)

**Reply by 204065248** [reply]
好吧，我利用https://codesandbox.io/s/inspiring-voice-zngl19?file=/src/index.ts这个代码实现了这个效果。但是无法调用command，请问有什么办法获得ctx呢

**Reply by Saul-Mirone** [reply]
```ts
import { $command } from '@milkdown/utils';
import { createCmd, createCmdKey } from '@milkdown/core';

type Options = {
  // some options
}

// 1. create a command key
export const MyCommandKey = createCmdKey<Options >('my-command-key');

// 2. implement the command
const myCommandPlugin = $command((ctx) => {
    return createCmd(MyCommandKey , (options: Options = {}) => {
      // your logic
    });
});

Editor.make()
// 3. use it here
.use(myCommandPlugin)
.use(menu.configure(menuPlugin, {
      config: [
        [{ type: "button", icon: "my-custom", key: "my-command-key"/* `MyCommandKey` also works */ }],
        ...defaultConfig
      ]
    })
).create();
```

**Reply by 204065248** [reply]
大佬理解错我的意思了。我是利用如下代码渲染成antd的dropdown组件
![image](https://user-images.githubusercontent.com/17307843/201055380-849afc65-12cd-4414-a0eb-577ceb61b3fe.png)
然后dropdown展示一个标签列表
![image](https://user-images.githubusercontent.com/17307843/201055527-c1128f34-c63f-46c7-b9db-c50044aaa64e.png)
在这个表情组件有一个点击的回调函数可以获得点击的表情是什么，可问题是此实无法获取ctx了。导致无法使用已经写好的command功能，目前我的两个思路是一个修改menu插件，另一个是把编辑器的变量存到全局【不过这样做会导致插件高耦合了】

**Reply by Saul-Mirone** [reply]
Maybe this?
```ts
const editor = await Editor.make().use(/*.....*/).create();

yourCallback((emojiString) => {
  editor.action((ctx) => {
    // add emoji to ctx
  });
});
```

**Reply by 204065248** [reply]
嗯，这种就是我想的第二个方案，将editor变量保存到全局，然后插件文件引入这个全局变量，就可以获得ctx。但是会造成高耦合

**Reply by Saul-Mirone** [reply]
I think it's fine. Or you can try this:

```ts
import type { Pre } from '@milkdown/core';

const myNord = () => (ctx) => nord.override((_, manager) => {
      const originalGet = manager.getSlice(ThemeIcon);

      manager.set(ThemeIcon, (icon) => {
        // you can get ctx now
      });
})(ctx as Pre)(ctx)
```

**Reply by 204065248** [reply]
了解，我先试试自己拓展menu插件，让他支持tsx组件。如果可以我就用那套方案了。实在不行就使用上述解决方案

**Reply by 204065248** [reply]
我把Menu菜单代码拉出来拓展了一下解决这个问题了

**Reply by Saul-Mirone** [reply]
👍 You can create a PR if it's a general solution.

**Reply by 204065248** [reply]
我这个解决方案是react才能实现的，纯js和vue暂时还没考虑过怎么实现，有点不通用
